### PR TITLE
OHE alignment

### PR DESCRIPTION
**Pull request by cometta**
Fixes  ISSUE #1150

New Features:
1.  FedProx
2. OHE

Codes are contributed by team mates:
https://github.com/jsweng  
https://github.com/nicholaslimt 
https://github.com/kwanchet

### CODE REVIEW

File: federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            values = [str(x) for x in pair_obj.values]            
+            col_maps[col_name] = values
+        
+        LOGGER.debug("**AIFEL** new col_maps is: {}".format(col_maps))
+
+        #Feature alignment is executed in a 'switch' manner
+        if self.need_alignment:
+
+            #Send col_maps to arbiter
+            self.transfer_variable.guest_columns.remote("{}".format(col_maps), role = consts.ARBITER, idx = -1) # send guest size to arbiter
```
Comment (tanmc123) [inline federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py:73]: FATE can transfer dict directly. Thus, I can't get why you transfer to string before remote.

File: federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                    continue
+                transform_col_names.append(col)
+                for vv in aligned_col_maps[col]:
+                    new_header.append(col+'_'+vv)
+
+            LOGGER.debug("**AIFEL** new transform col names after format received aligned columns: {}".format(transform_col_names))        
+            LOGGER.debug("**AIFEL** new header after format received aligned columns: {}".format(new_header))
+
+            self.inner_param.add_transform_names = transform_col_names
+            self.inner_param.set_result_header(new_header)
```
Comment (tanmc123) [inline federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py:99]: It seems this part is quite similar to the _transform_schema() func. May be you can consider reuse it.

File: federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            LOGGER.debug("**AIFEL** guest column sent")
+
+            #Receive aligned columns from arbiter
+            aligned_columns = self.transfer_variable.aligned_columns.get(idx=-1)
+            aligned_col_maps = ast.literal_eval(aligned_columns[0])
+            LOGGER.debug("**AIFEL** aligned columns received are: {}".format(aligned_col_maps))
+
+            ## transform with aligned columns
+            ## ['age', 'sex_0', 'sex_1', 'cp_1', 'cp_2', 'cp_3', 'cp_4', 'trestbps', 'chol', 'fbs_0', 'fbs_1', 'restecg_0', 'restecg_1', 'restecg_2', 'thalach', 'exang_0', 'exang_1', 'oldpeak', 'slope_1', 'slope_2', 'slope_3', 'ca_0', 'ca_1', 'ca_2', 'ca_3', 'thal_3', 'thal_6', 'thal_7']"
+
```
Comment (tanmc123) [inline federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py:83]: Hope those Logs and comments can be managered better. And some format is not quite friendly. Maybe a format brush can help.

File: federatedml/feature/OHE_with_alignment/OHE_alignment_host.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+from federatedml.util import consts
+import functools
+import ast
+from collections import defaultdict
+
+
+LOGGER = log_utils.getLogger()
+
+
+class OHEAlignmentHost(OHEAlignmentBase):
```
Comment (tanmc123) [inline federatedml/feature/OHE_with_alignment/OHE_alignment_host.py:23]: The logic of Host is almost same with Guest. Maybe you can create a "Client" class and make Guest and Host inherit it.

File: federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py (hunk 1 above)
Comment (cometta) [reply federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py]: refactor and removed this file

File: federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py (hunk 3 above)
Comment (cometta) [reply federatedml/feature/OHE_with_alignment/OHE_alignment_guest.py]: refactor and removed this file

File: federatedml/feature/OHE_with_alignment/OHE_alignment_host.py (hunk 4 above)
Comment (cometta) [reply federatedml/feature/OHE_with_alignment/OHE_alignment_host.py]: done refactored and deleted this file

**Pull request closed without merging**